imblearn/over_sampling/_smote_nc.py

```diff
@@ -0,0 +1,246 @@
+            warnings.warn('No "categorical_features" were specified when '
+                          'this instance was created. Will fall back '
+                          'to normal SMOTE', RuntimeWarning)
+            return super(SMOTENC, self)._fit_resample(X, y)
+
+        feature_indices = check_array(self.categorical_features, ensure_2d=False,
+                                      ensure_min_samples=2, estimator=self)
+        n_features = X.shape[1]
+
+        if np.any(feature_indices > n_features):
+            raise ValueError('Indices of categorical features have to be less '
+                             'than number of features in X: X.shape=(%s, %s)'
+                             % X.shape)
+
+        self.categorical_feature_indices_ = feature_indices
+        continuous_cols = np.delete(np.arange(n_features),
+                                    np.arange(feature_indices[0],
+                                              feature_indices[-1]))
+        if len(continuous_cols) == 0:
+            raise ValueError('Looks like all features in X are '
+                             'categorical which is not supported. '
+                             'For this method to work X should have '
+                             'at least 1 continuous feature.')
+
+        if sparse.issparse(X):
+            # let StandardScaler handle per-column variance in sparse matrix
+            scaler = StandardScaler(with_mean=False,
+                                    with_std=True,
+                                    copy=False)
+            scaler.fit(X.tocsc()[:, continuous_cols])
+            self.std_median_ = np.median(np.sqrt(scaler.var_))
+        else:
+            std = np.std(X[:, continuous_cols], axis=0)
+            self.std_median_ = np.median(std)
+
+        return super(SMOTENC, self)._fit_resample(X, y)
+
+    def _generate_sample(self, X, nn_data, nn_num, row, col, step):
+        """
+        Generates a new sample using normal SMOTE but then replaces
+        nominal features in generated sample according to SMOTE-NC:
+        with the value occurring in the majority of the k-nearest neighbors.
+        """
+        sample = super(SMOTENC, self)._generate_sample(X, nn_data, nn_num,
+                                                       row, col, step)
+        if not hasattr(self, "categorical_feature_indices_"):
+            warnings.warn('No "categorical_features" were specified when '
+                          'this instance was created. Will fall back '
+                          'to normal SMOTE', RuntimeWarning)
+            return sample
+
+        is_sparse = sparse.issparse(nn_data)
+        if is_sparse:
+            nn_data = nn_data.tocsr()
+
+        all_neighbors = nn_data[nn_num[row]]
+        if is_sparse:
+            all_neighbors = all_neighbors.tocsc()
+            sample = sample.tolil()
+
+        feature_idx_pairs = list(zip(self.categorical_feature_indices_[:-1],
```

Comment:
Would it be easier to use `itertools.combination`

```python
for start, end in combination(self.categorical_feature_indices_, 2):
    ...
```

Reply:
`feature_idx_pairs` here are *not* all possible pairs from the given iterable so itertools.combinations won't work. What we need are pairs made from two consecutive elements. Here's an example:

```python
categorical_feature_indices_ = [5, 8, 11, 15]
feature_idx_pairs = list(zip(categorical_feature_indices_[:-1], categorical_feature_indices_[1:]))
print(feature_idx_pairs)
# [(5, 8), (8, 11), (11, 15)]
```
I could also replace it like that:

```python
categorical_feature_indices_ = [5, 8, 11, 15]
for i, idx in enumerate(categorical_feature_indices_[1:]):
... start, end = categorical_feature_indices_[i], categorical_feature_indices_[i+1]
```